engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/MethodDefinitionExtractor.scala

```diff
@@ -96,7 +96,14 @@ private[definition] trait AbstractMethodDefinitionExtractor[T] extends MethodDef
       Option(method.getAnnotation(classOf[MethodToInvoke])).map(_.returnType())
       .filterNot(_ == classOf[Object])
       .map[TypingResult](Typed(_))
-    val typeFromSignature = EspTypeUtils.getGenericType(method.getGenericReturnType).map(Typed(_))
+    val typeFromSignature = {
+      val rawType = EspTypeUtils.getReturnClassForMethod(method)
+      (expectedReturnType, rawType) match {
+        // uwrap Future, Source and so on
```

Comment:
I wonder if we should have explicit list here? I don't think I understand the Source part - we don't handle it in inferGenericMonadType??

Reply:
IMO it not necessary - if someone has monad with one generic param in `expectedReturnType` we can assume that we are interested of this generic param. Adding explicit list here will add not necessary coupling. WDYT? Regards to Source, from what I saw in the code, we never handle type extraction correctly - if you haven't specified `returnType`, the type was always `Unknown`.

Reply:
Ok... let's leave it then... although I have some feeling that we have too many assumptions here...